### PR TITLE
[Object manager] don't abort entire pull request on race condition in concurrent chunk receive

### PR DESCRIPTION
**Pull request by mwtian**
## Why are these changes needed?
See #18062 for investigation and background.

This change ensures there is at most 1 inflight operation to create buffer for an object, when handling multiple chunks pushed from the object. This avoids the race condition where multiple operations race to create the buffer for the object and fail, forcing pulling to be retried.

Test from #18143 is pull into this change.

## Related issue number
#18062

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by ericl** [comment]
Instead of retrying at a higher level, which can increase load in 1->many broadcast situations, why not fix the race condition in the first place? E.g., if the chunk already exists (but is unsealed), don't raise an IOError (e.g., comment out the exception).

**Comment by rkooo567** [comment]
Yeah we should not reduce the retry interval. This will increase the load to systems a lot as our retry policy is pretty naive.

I think there are 2 approaches here. 

1. The Eric's suggestion. I think it is similar to my original approach https://github.com/ray-project/ray/pull/18143. 
2. We should retry within the pull manager. For example, in the HandlePull, instead of replying to other raylet right away, we cache the reply and retry creating the chunk in 1~2 seconds and reply after that. This will be simpler solution than 1 I believe

**Comment by mwtian** [comment]
Thanks @ericl and @rkooo567  for the suggestions. I went with a solution similar to [Eric's](https://github.com/ray-project/ray/pull/18955#issuecomment-929688910) and Sang's original [approach](https://github.com/ray-project/ray/pull/18143). The difference is instead of allowing multiple inflight create buffer ops, only 1 inflight create buffer ops is allowed. Because if we ignore already exists errors during buffer creation, there still needs to be a way for operations to wait until their corresponding `create_buffer_state_[object_id]` becomes available, which is more complicated than not having duplicated buffer creation requests.

The new logic can be different from the previous logic in thread and memory usage, under race condition. But that should be similar to most of the proposed solutions. If that is problematic, I can look into turning the logic async, or push flow control at receiving and sending sides.

File: src/ray/common/ray_config_def.h [hunk 1]
```diff
@@ -183,8 +183,7 @@ RAY_CONFIG(int64_t, worker_register_timeout_seconds, 30)
 RAY_CONFIG(int64_t, redis_db_connect_retries, 50)
 RAY_CONFIG(int64_t, redis_db_connect_wait_milliseconds, 100)
 
-/// Timeout, in milliseconds, to wait before retrying a failed pull in the
-/// ObjectManager.
+/// The object manager's global timer interval in milliseconds.
```
Comment (ericl) [inline src/ray/common/ray_config_def.h:186]: Please revert unrelated changes.

File: src/ray/object_manager/object_buffer_pool.cc [hunk 2]
```diff
@@ -179,6 +170,66 @@ std::vector<ObjectBufferPool::ChunkInfo> ObjectBufferPool::BuildChunks(
   return chunks;
 }
 
+ray::Status ObjectBufferPool::EnsureBufferExists(
+    const ObjectID &object_id, const rpc::Address &owner_address, uint64_t data_size,
+    uint64_t metadata_size, uint64_t chunk_index, std::unique_lock<std::mutex> &lock) {
+  RAY_CHECK(lock.owns_lock());
+
+  // Buffer for object_id already exists.
+  if (create_buffer_state_.contains(object_id)) return ray::Status::OK();
```
Comment (ericl) [inline src/ray/object_manager/object_buffer_pool.cc:179]: I think our style convention is to always put returns on a new line with braces, and never inline.

File: src/ray/object_manager/object_buffer_pool.cc [hunk 3]
```diff
@@ -179,6 +170,66 @@ std::vector<ObjectBufferPool::ChunkInfo> ObjectBufferPool::BuildChunks(
...
+
+  auto it = create_buffer_ops_.find(object_id);
+  if (it != create_buffer_ops_.end()) {
+    auto cond_var = it->second;
+    // Release lock while waiting, until there is no inflight create buffer operation.
+    cond_var->wait(
+        lock, [this, object_id]() { return !create_buffer_ops_.contains(object_id); });
+    // Buffer already created.
+    if (create_buffer_state_.contains(object_id)) return ray::Status::OK();
```
Comment (ericl) [inline src/ray/object_manager/object_buffer_pool.cc:188]: return on new line

File: src/ray/object_manager/object_buffer_pool.h [hunk 4]
```diff
@@ -166,13 +177,19 @@ class ObjectBufferPool {
...
   const ChunkInfo errored_chunk_ = {0, nullptr, 0, nullptr};
 
-  /// Mutex on public methods for thread-safe operations on
-  /// get_buffer_state_, create_buffer_state_, and store_client_.
-  mutable std::mutex pool_mutex_;
   /// Determines the maximum chunk size to be transferred by a single thread.
   const uint64_t default_chunk_size_;
+
+  /// Mutex to protect create_buffer_ops_ and create_buffer_state_.
+  mutable std::mutex pool_mutex_;
```
Comment (ericl) [inline src/ray/object_manager/object_buffer_pool.h:184]: Can you change this to use absl condition var support instead of std::condition_variable? https://abseil.io/docs/cpp/guides/synchronization

**Review by ericl — CHANGES_REQUESTED** [review]
cc @iycheng for a more detailed review

**Comment by ericl** [comment]
@rkooo567 , solution 2 isn't solving the fundamental issue. Solution 1 is simple, faster, and solves the root "bug" here.

File: src/ray/object_manager/object_buffer_pool.h [hunk 5]
```diff
@@ -119,34 +122,45 @@ class ObjectBufferPool {
...
+      ABSL_EXCLUSIVE_LOCKS_REQUIRED(pool_mutex_);
+
+  /// Ensures buffer for the object exists, and creates the buffer if needed.
+  /// Returns OK if buffer exists.
+  /// Must hold pool_mutex_ when calling this function. pool_mutex_ can be released
+  /// during the call.
+  ray::Status EnsureBufferExists(const ObjectID &object_id,
+                                 const rpc::Address &owner_address, uint64_t data_size,
+                                 uint64_t metadata_size, uint64_t chunk_index)
+      ABSL_EXCLUSIVE_LOCKS_REQUIRED(pool_mutex_);
```
Comment (ericl) [inline src/ray/object_manager/object_buffer_pool.h:158]: Btw I think we are omitting the `ABSL_` prefix on annotations.

**Review by ericl — COMMENTED** [review]
Looks good, but please change the annotations to be consistent with others (no ABSL_ prefix).

**Review by ericl — APPROVED** [review]

File: src/ray/object_manager/object_buffer_pool.h (hunk 5 above)
Comment (mwtian) [reply src/ray/object_manager/object_buffer_pool.h]: Good to know, updated.

**Comment by ericl** [comment]
Windows build broken in master
